### PR TITLE
PersonaData models

### PR DESCRIPTION
**Pull request by CyonAlexRDX**
* Add full impl of *models* of `PersonaData`
* Add `Personas`: collection of multiple Personas (like `Accounts`)
* Add `personas: Persona` to ProfileNetwork
* Make use of placeholder of Personas (with PersonaData) all the way to `Profile`
* With JSON roundtrip tests for all models as usual...
* Maintain code coverage of 99%.

Here is a `cloc` diff of this PR:

```sh
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Rust
 same                            0              0            142           2481
 modified                       12              0              1            400
 added                           9            193             19           2703
 removed                         0              0              1             21
```

As we can see, it is ~2700 loc of Rust added, still this is a false posive, since this includes the HUGE chunk of multiline string literals of JSON in Rust unit tests - counted as Rust code, but really is JSON. 

I plan to address this by moving these kinds of JSON values into actual `.json` files, like we have in integration tests already.

TL;DR; loc delta of actual Rust code is probably closer to 1000.

### CODE REVIEW

File: profile/src/v100/entity/persona/persona_data/entry_kinds/name.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        nickname: impl AsRef<str>,
+    ) -> Result<Self> {
+        let family_name = family_name.as_ref().trim().to_string();
+        let given_names = given_names.as_ref().trim().to_string();
+        let nickname = nickname.as_ref().trim().to_string();
+        if family_name.is_empty()
+            || given_names.is_empty()
+            || nickname.is_empty()
+        {
+            return Err(CommonError::InvalidDisplayNameEmpty);
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona_data/entry_kinds/name.rs]: wrong error type, create own, also three different, for granularity, for family name, given names and nickname.

File: profile/src/v100/entity/persona/persona_data/entry_kinds/name.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    Western,
+    Eastern,
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::prelude::*;
+
+    #[test]
+    fn new_name() {
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona_data/entry_kinds/name.rs]: add test which uses two given names as well

File: profile/src/v100/entity/persona/persona_data/entry_kinds/name.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+    #[test]
+    fn empty_given_name() {
+        let name = Name::new(Variant::Western, "Kent", "", "Superman");
+        assert_eq!(name, Err(CommonError::InvalidDisplayNameEmpty));
+    }
+
+    #[test]
+    fn empty_nickname() {
+        let name = Name::new(Variant::Western, "Kent", "Clark", "");
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona_data/entry_kinds/name.rs]: Ah actually nickname should be allowed to be empty

File: profile/src/v100/entity/persona/persona_data/entry_kinds/phone_number.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    Hash,
+    Eq,
+    derive_more::Display,
+    derive_more::Debug,
+    uniffi::Record,
+)]
+#[display("{number}")]
+#[debug("{number}")]
+#[serde(transparent)]
+pub struct PhoneNumber {
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona_data/entry_kinds/phone_number.rs]: reconsider macro merging `PhoneNumber` and `EmailAddress` (and possible `DisplayName`)???

File: profile/src/v100/entity/persona/persona_data/persona_data.rs [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    Hash,
+    Eq,
+    derive_more::Display,
+    derive_more::Debug,
+    uniffi::Record,
+)]
+#[display("{}", self.string_representation(false))]
+#[debug("{}", self.string_representation(true))]
+#[serde(rename_all = "camelCase")]
+pub struct PersonaData {
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona_data/persona_data.rs:37]: doc

File: profile/src/v100/entity/persona/persona_data/entry_kinds/phone_number.rs (hunk 4 above)
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona_data/entry_kinds/phone_number.rs]: doc

File: profile/src/v100/entity/persona/persona_data/entry_kinds/name.rs [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    Hash,
+    Eq,
+    derive_more::Display,
+    derive_more::Debug,
+    uniffi::Record,
+)]
+#[display("{}", self.full())]
+#[debug("{}", self.full())]
+#[serde(rename_all = "camelCase")]
+pub struct Name {
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona_data/entry_kinds/name.rs]: doc

File: profile/src/v100/entity/persona/persona_data/entry_kinds/email_address.rs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    Hash,
+    Eq,
+    derive_more::Display,
+    derive_more::Debug,
+    uniffi::Record,
+)]
+#[display("{email}")]
+#[debug("{email}")]
+#[serde(transparent)]
+pub struct EmailAddress {
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona_data/entry_kinds/email_address.rs]: doc

File: profile/src/v100/entity/persona/persona_data/persona_data_identified_collection_types.rs [hunk 8]
```diff
@@ -0,0 +1,349 @@
+use crate::prelude::*;
+
+macro_rules! declare_collection_of_identified_entry {
+    ($id_ent_type:ty,$struct_name:ident) => {
+        #[derive(
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona_data/persona_data_identified_collection_types.rs:5]: doc?

File: profile/src/v100/networks/network/personas.rs [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+            Persona::placeholder_stokenet_leia_skywalker(),
+            Persona::placeholder_stokenet_hermione(),
+        ])
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::prelude::*;
+    #[test]
```
Comment (Sajjon) [inline profile/src/v100/networks/network/personas.rs:81]: newline

File: profile/src/v100/networks/network/profile_network.rs [hunk 10]
```diff
@@ -40,6 +47,7 @@ impl ProfileNetwork {
         Self {
```
Comment (Sajjon) [inline profile/src/v100/networks/network/profile_network.rs:54]: Assert that all personas are on the same network and on the same network as accounts

File: profile/src/v100/networks/network/profile_network.rs [hunk 11]
```diff
@@ -136,17 +153,18 @@ mod tests {
...
                 Account::placeholder_stokenet(),
             ]),
+            Personas::default(),
         );
     }
 
     #[test]
-    fn json_roundtrip_placeholder_stokenet() {
-        let sut = ProfileNetwork::placeholder_stokenet();
+    fn json_roundtrip_placeholder_mainnet() {
```
Comment (Sajjon) [inline profile/src/v100/networks/network/profile_network.rs:180]: ah annoying git diff, I swapped `json_roundtrip_placeholder_mainnet()` and `json_roundtrip_placeholder_stokenet() `

File: profile/src/v100/entity/persona/persona.rs [hunk 12]
```diff
@@ -1,13 +1,23 @@
...
+    Debug,
+    PartialEq,
+    Hash,
+    Eq,
+    derive_more::Display,
+    uniffi::Record,
 )]
+#[display("{} | {}", display_name, address)]
 #[serde(rename_all = "camelCase")]
 pub struct Persona {
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona.rs:41]: doc

File: profile/src/v100/entity/persona/persona.rs [hunk 13]
```diff
@@ -58,49 +68,180 @@ impl Persona {
...
             mwp.clone(),
             WalletClientModel::Iphone,
         );
+
         let private_hd_factor_source =
             PrivateHierarchicalDeterministicFactorSource::new(mwp, bdfs);
-        let persona_creating_factor_instance: HDFactorInstanceTransactionSigning<IdentityPath> =
-            private_hd_factor_source.derive_entity_creation_factor_instance(network_id, index);
+
+        let ctr = Arc::<Mutex<u64>>::new(Mutex::new(1));
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona.rs:123]: nice pattern actually, hmm, generalize?

File: profile/src/v100/entity/persona/persona.rs [hunk 14]
```diff
@@ -58,49 +68,180 @@ impl Persona {
...
-            persona_data,
+            persona_data: persona_data.unwrap_or_default(),
         }
     }
+}
 
-    fn placeholder_at_index_name_network(
+impl Persona {
+    #[cfg(not(tarpaulin_include))] // false negative
+    fn placeholder_at_index_name_network<P, E>(
```
Comment (Sajjon) [inline profile/src/v100/entity/persona/persona.rs:100]: add flag `hidden` for some placeholder value? and to the same for account?

File: profile/src/v100/entity/persona/persona_data/entry_kinds/phone_number.rs (hunk 4 above)
Comment (Sajjon) [reply profile/src/v100/entity/persona/persona_data/entry_kinds/phone_number.rs]: hmm I deem it just not worth it. In future we might wanna add different validation, and also it makes documentation a bit harder since harder to document macro defined structs.

**Comment by Sajjon** [comment]
Merge as is, since I need to start AuthorizedDapps models which requires this, I will address any comments in "late" review by Ghenadie in the AuthorizedDapps working branch

File: profile/src/v100/entity/persona/persona_data/persona_data_identified_collection_types.rs [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+            }
+
+            fn placeholder_other() -> Self {
+                $struct_name::new(<$id_ent_type>::placeholder_other())
+            }
+        }
+    };
+}
+
+declare_collection_of_identified_entry!(
```
Comment (GhenadieVP) [inline profile/src/v100/entity/persona/persona_data/persona_data_identified_collection_types.rs:92]: nice usage of macros, took might some minutes to figure out what you did, but this is awesome :)!

**Review by GhenadieVP — COMMENTED** [review]
No comments, awesome work 💯!
